RoBERTa/fairseq.patch

```diff
@@ -162,6 +162,71 @@ index da1f9491..ec6e1732 100644
          self.pipeline_model_parallel = cfg.distributed_training.pipeline_model_parallel
          self.last_device = None
          if self.cuda and self.pipeline_model_parallel:
+diff --git a/fairseq_cli/train.py b/fairseq_cli/train.py
+index 376bd1d0..2b10d49c 100644
+--- a/fairseq_cli/train.py
++++ b/fairseq_cli/train.py
+@@ -26,6 +26,7 @@ logger = logging.getLogger("fairseq_cli.train")
+ import numpy as np
+ import torch
+ from omegaconf import DictConfig, OmegaConf
++from fvcore.nn import FlopCountAnalysis
+ 
+ from fairseq import checkpoint_utils, options, quantization_utils, tasks, utils
+ from fairseq.data import data_utils, iterators
+@@ -123,6 +124,21 @@ def main(cfg: FairseqConfig) -> None:
+         )
+     )
+ 
++    # compute model flops: create fake input and use fvcore to compute the model flops.
++    tokens_per_sample = cfg.task.tokens_per_sample
++    logger.info(f'tokens_per_sample: {tokens_per_sample}')
++    token_ids = []
++    token_ids.append(0)
++
++    for _ in range(tokens_per_sample - 2):
++        token_ids.append(100)
++    token_ids.append(2)
++    src_tokens = torch.tensor(token_ids).unsqueeze(0).cuda()
++    with torch.no_grad():
++        model.cuda()
++        model_flops = FlopCountAnalysis(model, src_tokens).total()
++    logger.info(f'model flops: {model_flops}')
++
+     # Load valid dataset (we load training data below, based on the latest checkpoint)
+     # We load the valid dataset AFTER building the model
+     data_utils.raise_if_valid_subsets_unintentionally_ignored(cfg)
+@@ -187,7 +203,7 @@ def main(cfg: FairseqConfig) -> None:
+             break
+ 
+         # train for one epoch
+-        valid_losses, should_stop = train(cfg, trainer, task, epoch_itr)
++        valid_losses, should_stop = train(cfg, trainer, task, epoch_itr, model_flops)
+         if should_stop:
+             break
+ 
+@@ -244,7 +260,7 @@ def should_stop_early(cfg: DictConfig, valid_loss: float) -> bool:
+ 
+ @metrics.aggregate("train")
+ def train(
+-    cfg: DictConfig, trainer: Trainer, task: tasks.FairseqTask, epoch_itr
++    cfg: DictConfig, trainer: Trainer, task: tasks.FairseqTask, epoch_itr, model_flops=0
+ ) -> Tuple[List[Optional[float]], bool]:
+     """Train the model for one epoch and return validation losses."""
+     # Initialize data iterator
+@@ -320,6 +336,12 @@ def train(
+             num_updates = trainer.get_num_updates()
+             if num_updates % cfg.common.log_interval == 0:
+                 stats = get_training_stats(metrics.get_smoothed_values("train_inner"))
++                throughput = stats["bsz"] * stats["ups"]
++                stats["throughput"] = throughput
++                throughput_per_gpu = throughput / cfg.distributed_training.distributed_world_size
++                # The reason of x 3: 1 forward + 2 backward. The reason of x 2: 1MACC = 2FLOPS
```

Comment:
Do you know why 1MAC = 2FLOPS?

Reply:
> Do you know why 1MAC = 2FLOPS?

1 MAC contains an addition operator and a multiplication operator.
In the unit FLOP, an addition operator is 1 FLOP, and a multiplication operator is also 1 FLOP.

Reply:
Got it. Thanks.